plutus-core/plutus-core/src/PlutusCore/DeBruijn.hs

```diff
@@ -112,72 +119,125 @@ deBruijnTermM = \case
     Builtin ann bn -> pure $ Builtin ann bn
 
 -- | Convert a 'Type' with 'NamedTyDeBruijn's into a 'Type' with 'TyName's.
+-- Will throw an error if a free variable is encountered.
 unDeBruijnTy
     :: (MonadQuote m, AsFreeVariableError e, MonadError e m)
     => Type NamedTyDeBruijn uni ann -> m (Type TyName uni ann)
 unDeBruijnTy = flip runReaderT (Levels 0 BM.empty) . unDeBruijnTyM
 
 -- | Convert a 'Term' with 'NamedTyDeBruijn's and 'NamedDeBruijn's into a 'Term' with 'TyName's and 'Name's.
+-- Will throw an error if a free variable is encountered.
 unDeBruijnTerm
     :: (MonadQuote m, AsFreeVariableError e, MonadError e m)
     => Term NamedTyDeBruijn NamedDeBruijn uni fun ann -> m (Term TyName Name uni fun ann)
 unDeBruijnTerm = flip runReaderT (Levels 0 BM.empty) . unDeBruijnTermM
 
 -- | Convert a 'Program' with 'NamedTyDeBruijn's and 'NamedDeBruijn's into a 'Program' with 'TyName's and 'Name's.
+-- Will throw an error if a free variable is encountered.
 unDeBruijnProgram
     :: (MonadQuote m, AsFreeVariableError e, MonadError e m)
     => Program NamedTyDeBruijn NamedDeBruijn uni fun ann -> m (Program TyName Name uni fun ann)
 unDeBruijnProgram (Program ann ver term) = Program ann ver <$> unDeBruijnTerm term
 
+
+-- | Convert a 'Type' with 'NamedTyDeBruijn's into a 'Type' with 'TyName's.
+-- A variant that does not throw an error on free variables. See 'freeIndexGrace'.
+unDeBruijnTyGrace
```

Comment:
Suggestion: don't define/export these, but rather export the `*With` variants and also the graceful handler. It's not so bad for the caller to specify the handler in the unusual cases where you want this behaviour, it'll probably even be less code. I'd even be tempted not to export the existing ones.... but I guess they are normally what you want.

Reply:
Ok, but it is the responsibility of the "users" of the api to initiall setup the cache of the grace mechanism (evalState M.empty)